opm/grid/cpgrid/CpGrid.cpp

```diff
@@ -2206,159 +2204,122 @@ void CpGrid::addLgrsUpdateLeafView(const std::vector<std::array<int,3>>& cells_p
-                }
+        // Predict how many new cells/points (born in refined level grids) need new globalIds, so we can assign uniquily
+        // new ids ( and anticipat the maximum).
+        // At this point, neither cell_index_set_ nor partition_type_indicator_ are populated.
+        // Refined level grid cells:
+        //    1. Inherit their partition type from their parent cell (i.e., element.father().partitionType()).
+        //    2. Assign global ids only for interior cells.
+        // Refined level grid points/vertices:
+        //    1. Challenge: <refined level grid>.computePointPartitionType(). This requires face and cell partition type to be defined. For
+        //       a grid with LGRs, this can be done on the fly. Under the assumption of fully interior LGRs, all refined level grid
+        //       points are interior.
+        //    2. Previous step allows us to assign uniquily global ids to non-overlap points.
+        // Under the assumption of LGRs fully-interior, no communication is needed. In the general case, communication will be used
+        // to populate overlap cell/point global ids on the refined level grids.
+
+        std::vector<std::size_t> cell_ids_needed_by_proc(comm().size());
+        std::size_t local_cell_ids_needed = 0;
+        for ( const auto& element : elements( levelGridView(0), Dune::Partitions::interior) ) {
+            // Get old mark (from level zero). After calling adapt, all marks are set to zero.
+            bool hasBeenMarked = currentData().front()->getMark(element) == 1;
+            if ( hasBeenMarked ) {
+                const auto& level = assignRefinedLevel[element.index()];
+                local_cell_ids_needed += cells_per_dim_vec[level-1][0]*cells_per_dim_vec[level-1][1]*cells_per_dim_vec[level-1][2];
             }
-            global_cells_per_level[level] = comm().sum(local_owned_cells_per_level[level]);
-            if (level)
-            {
-                for (const auto& point : vertices(levelGridView(level))) {
-                    if (point.partitionType() == InteriorEntity) {
-                        const auto& bornLevel =  (*current_data_)[level]->corner_history_[point.index()][0];
-                        if(bornLevel == -1) { // It means that the corner is a new born one, i.e. does not
-                            // coincide with any corner from level zero.
-                            ++local_owned_points_per_level[level];
+        }
+        comm().allgather(&local_cell_ids_needed, 1, cell_ids_needed_by_proc.data());
+
+        std::vector<std::size_t> point_ids_needed_by_proc(comm().size());
+        std::size_t local_point_ids_needed = 0;
+        for (std::size_t level = 1; level < cells_per_dim_vec.size()+1; ++level){
+            // Compute the partition type for point
+            if(active_cell_count_per_level[level-1]>0) {
+                (*current_data_)[level]->computePointPartitionType();
```

Comment:
Maybe put comment here that a TODO is still to set up the parallel index set correctly?
Currently this call will mark all elements as owner. I guess that is probably expected as we don't allow LGRs touching the faces of overlap cells, yet?

Reply:
I will add a TODO comment, but it's sad to realize that it’s not working as I expected. I initially thought that computePointPartition() would categorize points as interior, border, front, or overlap based on [CpGridData::computePointPartition() ](https://github.com/OPM/opm-grid/blob/master/opm/grid/cpgrid/CpGridData.cpp#L1631)